### PR TITLE
Admin endpoints cleanup

### PR DESCRIPTION
**Pull request by Quetzacoalt91**
:warning: Assets not generated

| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | This PR provides a controller-equivalent method to the different request made outside the `index.php` files
| Type?         | new feature
| Category?     | CO
| BC breaks?    | Yes, entrypoints now require a token to be provided.
| Deprecations? | Yes, all entrypoints expect `index.php` will be deprecated.
| Fixed ticket? | https://github.com/PrestaShop/PrestaShop/issues/10063
| How to test?  | All request must pass on the new and old URLs, and the URLs used should all target a `index.php` file.

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10131)

### CODE REVIEW

File: controllers/admin/AdminBackupController.php [hunk 1]
```diff
@@ -318,4 +318,83 @@ public function strSort($a, $b)
         return $this->_orderWay == 'ASC' ? strcmp($a[$this->sort_by], $b[$this->sort_by]) :
             strcmp($b[$this->sort_by], $a[$this->sort_by]);
     }
+
+    /**
+     * Reads and output the content of a backup file
+     *
+     * Expected query parameter {filename}
+     */
+    public function displayAjaxBackupContent()
```
Comment (mickaelandrieu) [inline controllers/admin/AdminBackupController.php]: You need to rebase, we have migrated this controller to Symfony

Comment (Quetzacoalt91) [reply controllers/admin/AdminBackupController.php]: hmm I will probably need to revert a commit if we want to avoid a BC break.

**Comment by PierreRambaud** [comment]
61 files, still in WIP, the review will be really fun :trollface:

**Comment by Quetzacoalt91** [comment]
Not in progress anymore, one rebase and you're good to go :)

File: classes/PrestaShopBackup.php [hunk 2]
```diff
@@ -155,10 +155,13 @@ public static function backupExist($filename)
...
      *
      * @return string The url used to request the backup file
+     *
+     * @deprecated As the call has been duplicated in the new Controller. Get the URL from the router instead.
      */
     public function getBackupURL()
     {
-        return __PS_BASE_URI__ . basename(_PS_ADMIN_DIR_) . '/backup.php?filename=' . basename($this->id);
+        // Additionnal parameters (action, filename, ajax) are kept for backward compatibility, in case we disable the new controller
+        return Context::getContext()->link->getAdminLink('AdminBackup', true, ['route' => 'admin_backup_download', 'downloadFileName' => basename($this->id)]) . '&action=backupContent&ajax=1&filename=' . basename($this->id);
```
Comment (matks) [inline classes/PrestaShopBackup.php]: @PierreRambaud will tell you that this line is too long

File: classes/module/ModuleGraph.php [hunk 3]
```diff
@@ -305,7 +305,9 @@ public function engine($params)
         $url_params['module'] = Tools::getValue('module');
         $url_params['id_employee'] = $id_employee;
         $url_params['id_lang'] = $id_lang;
-        $drawer = 'drawer.php?' . http_build_query(array_map('Tools::safeOutput', $url_params), '', '&');
+        $url_params['ajax'] = 1;
+        $url_params['action'] = 'graphDraw';
+        $drawer = Context::getContext()->link->getAdminLink('AdminStats') . '&' . http_build_query(array_map('Tools::safeOutput', $url_params), '', '&');
```
Comment (matks) [inline classes/module/ModuleGraph.php]: "Line too long"
 \- GoT

File: controllers/admin/AdminCategoriesController.php [hunk 4]
```diff
@@ -1101,4 +1101,65 @@ public function ajaxProcessuploadThumbnailImages()
...
         }
     }
+
+    /**
+     * Display children categories.
+     *
+     * Required request params:
+     * - id_category_parent
+     * - selectedCat
+     * -use_shop_context
```
Comment (matks) [inline controllers/admin/AdminCategoriesController.php]: Missing a white space `* - use_shop_context`

File: controllers/admin/AdminProductsController.php [hunk 5]
```diff
@@ -3214,4 +3214,157 @@ public function processImageLegends()
...
+            SELECT p.`id_product`, pl.`name`
+            FROM `' . _DB_PREFIX_ . 'product` p
+            NATURAL LEFT JOIN `' . _DB_PREFIX_ . 'product_lang` pl
+            WHERE pl.`id_lang` = ' . (int) (Tools::getValue('id_lang')) . '
+            ' . Shop::addSqlRestrictionOnLang('pl') . '
+            AND NOT EXISTS (SELECT 1 FROM `' . _DB_PREFIX_ . 'pack` WHERE `id_product_pack` = p.`id_product`)
+            AND p.`id_product` != ' . (int) (Tools::getValue('id_product')));
+
+        foreach ($products as $packItem) {
+            $jsonArray[] = '{"value": "' . (int) ($packItem['id_product']) . '-' . addslashes($packItem['name']) . '", "text":"' . (int) ($packItem['id_product']) . ' - ' . addslashes($packItem['name']) . '"}';
```
Comment (matks) [inline controllers/admin/AdminProductsController.php]: This looks like a good target for a refacto into a private function

**Review by tomlev — DISMISSED** [review]
> Reviewed 16 of 40 files at r1, 1 of 17 files at r2, 43 of 43 files at r3.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10131)* status: all files reviewed, 11 unresolved discussions (waiting on @mickaelandrieu, @matks, and @Quetzacoalt91)

___
*[admin-dev/ajax.php, line 153 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLcpIo7ACX50lcoOV0C:-LLcpIo7ACX50lcoOV0D:bh5w0yc) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/admin-dev/ajax.php#L153)):*
> ```PHP
> if (Tools::isSubmit('getNotifications')) {
>     $notification = new Notification;
>     die(json_encode($notification->getLastElements()));
> ```

you should echo the result then die(0) to be compliant with CLI calls
___
*[controllers/admin/AdminProductsController.php, line 3246 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLd5rfuDGihD7KQyjC1:-LLd5rfuDGihD7KQyjC2:bndqcme) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/controllers/admin/AdminProductsController.php#L3246)):*
> ```PHP
>     {
>         $query = Tools::getValue('q', false);
>         if (!$query || $query == '' || strlen($query) < 1) {
> ```

why not using "if (empty($query)) {" ?
___
*[controllers/admin/AdminSearchController.php, line 43 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLdll9_ARp2yCvBBoVj:-LLdll9_ARp2yCvBBoVk:bur2p3) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/controllers/admin/AdminSearchController.php#L43)):*
> ```PHP
>         // Specific check for the ajax request 'searchCron'
>         if (Tools::isSubmit('action') && 'searchCron' === Tools::getValue('action')) {
>             if (substr(_COOKIE_KEY_, 34, 8) === Tools::getValue('token')) {
> ```

is there any way here to set a constant instead of using magic numbers ?
___
*[controllers/admin/AdminSearchController.php, line 492 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLdlvCh2oai7jN_GSsJ:-LLdlvCh2oai7jN_GSsK:b3clfhc) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/controllers/admin/AdminSearchController.php#L492)):*
> ```PHP
>         }
> 
>         ini_set('max_execution_time', 7200);
> ```

why 2 hours ? would you please provide some reason in comments to help understanding this arbitrary duration ?
___
*[controllers/admin/AdminZonesController.php, line 149 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLdmsxJChn7f0RU1KYX:-LLdmsxJChn7f0RU1KYY:bkfs71y) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/controllers/admin/AdminZonesController.php#L149)):*
> ```PHP
>     public function displayAjaxZones()
>     {
>         $html = '<select id="zone_to_affect" name="zone_to_affect">';
> ```

it could be usefull to move it into a template file
___
*[src/PrestaShopBundle/Entity/ProductDownload.php, line 281 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLdxecD8AVoynfVRq-B:-LLdxecD8AVoynfVRq-C:b-1ljw13) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/src/PrestaShopBundle/Entity/ProductDownload.php#L281)):*
> ```PHP
>     {
>         if ($this->getDateAdd() == null) {
>             $this->setDateAdd(new DateTime(date('Y-m-d H:i:s')));
> ```

only "new DateTime()" is enough

File: controllers/admin/AdminProductsController.php (hunk 5 above)
Comment (Quetzacoalt91) [reply controllers/admin/AdminProductsController.php]: What would you see in this private method?

**Comment by Quetzacoalt91** [comment]
Thank you @matks & @tomlev, I just updated the PR with your comments.

**Comment by mickaelandrieu** [comment]
Hmm, sounds like risky contribution and I'm worried about the BC break part: as I agree with the objective of this pull request, I don't see the gain for PrestaShop users: we will probably create issues and they won't see any gains.

Can we agree on postponing that to 1.7.6 or does I miss something critical here? /c @Quetzacoalt91

**Comment by eternoendless** [comment]
Ok to postpone until 1.7.6

File: admin-dev/themes/default/template/controllers/orders/_customized_data.tpl [hunk 6]
```diff
@@ -129,7 +129,7 @@
 									<div class="form-group">
 										<span class="col-lg-4 control-label"><strong>{if $data['name']}{$data['name']}{else}{l s='Picture #' d='Admin.Orderscustomers.Feature'}{$data@iteration}{/if}</strong></span>
 										<div class="col-lg-8">
-											<a href="displayImage.php?img={$data['value']}&amp;name={$order->id|intval}-file{$data@iteration}" class="_blank">
+											<a href="{$link->getAdminLink('AdminCarts')}&ajax=1&action=customizationImage&img={$data['value']}&amp;name={$order->id|intval}-file{$data@iteration}" class="_blank">
```
Comment (PierreRambaud) [inline admin-dev/themes/default/template/controllers/orders/_customized_data.tpl:132]: We need to use the full syntax, wdyt? (Maybe in an other PR)
```
{$link->getAdminLink('AdminCarts', true ['ajax' => 1, 'action' => 'customizationImage', 'img' => $data['value'], 'name' => $order->id|intval . '-file' . $data@iteration])}
```

File: classes/Referrer.php [hunk 7]
```diff
@@ -387,6 +387,6 @@ public static function getAjaxProduct($idReferrer, $idProduct, $idEmployee = nul
             'percent_fee' => Tools::displayPrice($statsSales['sales'] * $referrer->percent_fee / 100, $currency),
         );
 
-        die('[' . json_encode($jsonArray) . ']');
+        return '[' . json_encode($jsonArray) . ']';
```
Comment (PierreRambaud) [inline classes/Referrer.php:390]: `return json_encode([$jsonArray]);`

File: controllers/admin/AdminProductsController.php [hunk 8]
```diff
@@ -3214,4 +3214,158 @@ public function processImageLegends()
...
+                        'id' => $item['id_product'],
+                        'name' => $item['name'] . (!empty($item['reference']) ? ' (ref: ' . $item['reference'] . ')' : ''),
+                        'ref' => (!empty($item['reference']) ? $item['reference'] : ''),
+                        'image' => str_replace('http://', Tools::getShopProtocol(), $context->link->getImageLink($item['link_rewrite'], $item['id_image'], 'home_default')),
+                    );
+                }
+            }
+
+            if (!$forceJson) {
+                $this->ajaxRender(implode("\n", $results));
```
Comment (PierreRambaud) [inline controllers/admin/AdminProductsController.php:3331]: PHP_EOL instead of \n

File: controllers/admin/AdminReferrersController.php [hunk 9]
```diff
@@ -399,6 +370,34 @@ public function renderForm()
...
+        $result = Db::getInstance()->executeS('
+            SELECT p.id_product, pl.name
+            FROM ' . _DB_PREFIX_ . 'product p
+            LEFT JOIN ' . _DB_PREFIX_ . 'product_lang pl
+                ON (p.id_product = pl.id_product AND pl.id_lang = ' . (int) Tools::getValue('id_lang') . ')
+            ' . (Tools::getValue('filter') != 'undefined' ? 'WHERE name LIKE "%' . pSQL(Tools::getValue('filter')) . '%"' : '')
+        );
+
+        foreach ($result as $row) {
+            $json_array[] = '{id_product:' . (int) $row['id_product'] . ',name:\'' . addslashes($row['name']) . '\'}';
```
Comment (PierreRambaud) [inline controllers/admin/AdminReferrersController.php:395]:
```php
$json_array[] = [
  'id_product' => (int) $row['id_product'], 
  'name' => $row['name'],
];
```
and use json_encode in ajaxRender

File: controllers/admin/AdminTranslationsController.php [hunk 10]
```diff
@@ -3305,4 +3305,14 @@ public static function getEmailHTML($email)
...
     }
+
+    /**
+     * Display the HTML content of an email.
+     */
+    public function displayAjaxEmailHTML()
+    {
+        $email = Tools::getValue('email');
+        $this->ajaxRender(
+            AdminTranslationsController::getEmailHTML($email));
```
Comment (PierreRambaud) [inline controllers/admin/AdminTranslationsController.php:3333]: Weird indent

File: src/PrestaShopBundle/Controller/Admin/CommonController.php [hunk 11]
```diff
@@ -42,6 +42,33 @@
...
+
+    /**
+     * Update the last time a notification type has been seen.
+     *
+     * @param Request $request
+     */
+    public function notificationsAckAction(Request $request)
+    {
+        $type = $request->request->get('type');
+        // Adapter needed here?
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Controller/Admin/CommonController.php:68]: Ping @mickaelandrieu @matks 
I think an adapter is needed

File: src/PrestaShopBundle/Entity/ModuleHistory.php [hunk 12]
```diff
@@ -189,10 +189,10 @@ public function getDateUpd()
      */
     public function updatedTimestamps()
     {
-        $this->setDateUpd(new DateTime(date('Y-m-d H:i:s')));
+        $this->setDateUpd(new DateTime());
```
Comment (PierreRambaud) [inline src/PrestaShopBundle/Entity/ModuleHistory.php:192]: :+1:

File: src/PrestaShopBundle/Controller/Admin/CommonController.php (hunk 11 above)
Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/CommonController.php]: I'll create an issue in Migration kanban for this one. Let's not fix it in this PR.

Comment (matks) [reply src/PrestaShopBundle/Controller/Admin/CommonController.php:68]: Why us 😛 everybody can write an adapter

**Review by eternoendless — CHANGES_REQUESTED** [review]
How about adding no regression tests for this change?

1. Execute a query to the old endpoint, then save the result
2. Create a test that executes the request, then verifies the result against what you saved before
3. Move code around
4. Run tests... everything should be working the same as before :)

> Reviewed 1 of 40 files at r1, 1 of 17 files at r2, 1 of 43 files at r3, 13 of 30 files at r4, 27 of 27 files at r5.
> *[Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10131#-:-LVDo_BdCQm67zpLUp0T:bojw4kg)* status: all files reviewed, 36 unresolved discussions (waiting on @mickaelandrieu, @matks, @Quetzacoalt91, @tomlev, and @PierreRambaud)

___
*[admin-dev/ajax.php, line 43 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDiQNZ5Vcf47cPWVFb:-LVDiQNZ5Vcf47cPWVFc:bmb5dtf) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/admin-dev/ajax.php#L43)):*
> ```PHP
> /**
>  * Ajax calls to the controller AdminReferrers
>  * -> Moved in legacy
> ```

```diff
- Moved in legacy
+ Moved to legacy
```

(to be fixed everywhere :grin: )
___
*[admin-dev/ajax.php, line 53 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDi6OH70euE8-Hic7r:-LVDi6OI7roh_F1pTw8G:b-lxobdk) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/admin-dev/ajax.php#L53)):*
> ```PHP
>     if (Tools::isSubmit('ajaxFillProducts')) {
>         $_GET['action'] = 'fillProducts';
>     }
> ```

Instead of keeping this behavior, why not issue 301 moved permanently redirects to the correct URL?
___
*[admin-dev/themes/default/template/controllers/carts/helpers/view/view.tpl, line 122 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDt4jaEEdZgEsYOaZS:-LVDt4jaEEdZgEsYOaZT:bbintu8) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/admin-dev/themes/default/template/controllers/carts/helpers/view/view.tpl#L122)):*
> ```Smarty
> 									    {foreach from=$datas key='index' item='data'}
> 											    <li style="display: inline; margin: 2px;">
> 												    <a href="{$link->getAdminLink('AdminCarts')}&ajax=1&action=customizationImage&img={$data.value}&name={$order->id|intval}-file{$index}" class="_blank">
> ```

Remember to use the 4th argument of getAdminLink
___
*[admin-dev/themes/default/template/controllers/orders/_customized_data.tpl, line 132 at r4](https://reviewable.io/reviews/prestashop/prestashop/10131#-LKghrLjn47vzYlrLpMt-r4-132:-LVDt9Lw5ACEmE93MJoZ:b32v57) ([raw file](https://github.com/prestashop/prestashop/blob/bd4a26adadb6bd1b035073fe3b3631bf73548ead/admin-dev/themes/default/template/controllers/orders/_customized_data.tpl#L132)):*
<details><summary><i>Previously, PierreRambaud (GoT) wrote…</i></summary><blockquote>

We need to use the full syntax, wdyt? (Maybe in an other PR)
```
{$link->getAdminLink('AdminCarts', true ['ajax' => 1, 'action' => 'customizationImage', 'img' => $data['value'], 'name' => $order->id|intval . '-file' . $data@iteration])}
```
</blockquote></details>

+1
___
*[admin-dev/themes/default/template/controllers/return/helpers/form/form.tpl, line 85 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDkm-g2ot-v31IseuU:-LVDkm-g2ot-v31IseuV:bt7xuei) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/admin-dev/themes/default/template/controllers/return/helpers/form/form.tpl#L85)):*
> ```Smarty
> 												<span class="col-lg-3 control-label"><strong>{l s='Attachment'}</strong></span>
> 												<div class="col-lg-9">
> 													<a href="{$link->getAdminLink('AdminCarts')}&ajax=1&action=customizationImage&img={$data['value']}&amp;name={$returnedCustomization['id_order_detail']|intval}-file{$smarty.foreach.data.iteration.iteration}" class="_blank"></a>
> ```

You should use the 4th parameter of getAdminLink here (see [Automated link generation](https://devdocs.prestashop.com/1.7/modules/core_updates/1.7.5/#automated-link-generation))
___
*[classes/PrestaShopBackup.php, line 162 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDmxlaE8K1boPyfwh2:-LVDmxlaE8K1boPyfwh3:b1kek4c) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/classes/PrestaShopBackup.php#L162)):*
> ```PHP
>      * @return string The url used to request the backup file
>      *
>      * @deprecated As the call has been duplicated in the new Controller. Get the URL from the router instead.
> ```

`@deprecated Since 1.7.6.0 ...`
___
*[classes/PrestaShopBackup.php, line 175 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDmnwj4UCSg2ZiQK6E:-LVDmnwj4UCSg2ZiQK6F:b36b5a8) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/classes/PrestaShopBackup.php#L175)):*
> ```PHP
>                 ]
>             )
>             . '&action=backupContent&ajax=1&filename='
> ```

Use the 4th argument of getAdminLink instead
___
*[classes/Referrer.php, line 390 at r4](https://reviewable.io/reviews/prestashop/prestashop/10131#-LKgLbidp0fJF9AOABCP-r4-390:-LVDnLeW4_PmvnI8NybH:b32v57) ([raw file](https://github.com/prestashop/prestashop/blob/bd4a26adadb6bd1b035073fe3b3631bf73548ead/classes/Referrer.php#L390)):*
<details><summary><i>Previously, PierreRambaud (GoT) wrote…</i></summary><blockquote>

`return json_encode([$jsonArray]);`
</blockquote></details>

+1
___
*[classes/module/ModuleGraph.php, line 311 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDnTvCE2RYjlEM6TBL:-LVDnTvCE2RYjlEM6TBM:b36b5a8) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/classes/module/ModuleGraph.php#L311)):*
> ```PHP
>         $url_params['action'] = 'graphDraw';
>         $drawer = Context::getContext()->link->getAdminLink('AdminStats')
>             . '&' . http_build_query(array_map('Tools::safeOutput', $url_params), '', '&');
> ```

Use the 4th argument of getAdminLink instead
___
*[controllers/admin/AdminCartsController.php, line 852 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDns4tCaPCc1LD1Sd3:-LVDns4tCaPCc1LD1Sd4:bigzr6x) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/controllers/admin/AdminCartsController.php#L852)):*
> ```PHP
>         $img = Tools::getValue('img');
>         $name = Tools::getValue('name');
>         $path = _PS_UPLOAD_DIR_ . $img;
> ```

`$img` should be verified to avoid path traversal.
___
*[controllers/admin/AdminCategoriesController.php, line 1126 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDoAeVB_jW-SY4SQ8U:-LVDoAeVB_jW-SY4SQ8V:bhzl0lc) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/controllers/admin/AdminCategoriesController.php#L1126)):*
> ```PHP
>         $children_categories = Category::getChildrenWithNbSelectedSubCat(
>             Tools::getValue('id_category_parent'),
>             Tools::getValue('selectedCat'), Context::getContext()->language->id,
> ```

Watch out, there are two arguments on the same line here
___
*[controllers/admin/AdminProductsController.php, line 3234 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LKgLbidp0fJF9AOABCT-r3-3234:-LVDpDzt7OL5hg41vpKn:b4ykbl8) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/controllers/admin/AdminProductsController.php#L3234)):*
<details><summary><i>Previously, Quetzacoalt91 (Thomas N) wrote…</i></summary><blockquote>

What would you see in this private method?
</blockquote></details>

This can be refactored easily to avoid writing json as a string:

```php
foreach ($products as $packItem) {
    $response[] = [
        'value' => sprintf('%d-%s', $packItem['id_product'], $packItem['name']),
        'text' => sprintf('%d - %s', $packItem['id_product'], $packItem['name']),
    ];
}

return $this->ajaxRender(json_encode($response));
```
___
*[controllers/admin/AdminSearchConfController.php, line 67 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDqoec18Pui9Xm04hV:-LVDqoec18Pui9Xm04hW:bbintu8) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/controllers/admin/AdminSearchConfController.php#L67)):*
> ```PHP
> 
>         // Search options
>         $cron_url = Context::getContext()->link->getAdminLink('AdminSearch', false) . '&action=searchCron&ajax=1' .
> ```

Remember to use the 4th argument of getAdminLink
___
*[controllers/admin/AdminSearchConfController.php, line 84 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDqrXPDuSE7f_Rc63a:-LVDqrXPDuSE7f_Rc63b:bqujbyh) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/controllers/admin/AdminSearchConfController.php#L84)):*
> ```PHP
> 						' . $this->trans('If your server stops before the process ends, you can resume the indexing by clicking "Add missing products to the index".', array(), 'Admin.Shopparameters.Feature') . '
> 					</p>
> 					<a href="' . Context::getContext()->link->getAdminLink('AdminSearch', false) . '&action=searchCron&ajax=1&token=' . $this->getTokenForCron() . '&amp;redirect=1' . (Shop::getContext() == Shop::CONTEXT_SHOP ? '&id_shop=' . (int) Context::getContext()->shop->id : '') . '" class="btn-link">
> ```

Here too
___
*[controllers/admin/AdminSearchConfController.php, line 88 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDqshkA2qVK37jUWB1:-LVDqshkA2qVK37jUWB2:b-fulkv0) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/controllers/admin/AdminSearchConfController.php#L88)):*
> ```PHP
> 						' . $this->trans('Add missing products to the index', array(), 'Admin.Shopparameters.Feature') . '
> 					</a><br />
> 					<a href="' . Context::getContext()->link->getAdminLink('AdminSearch', false) . '&action=searchCron&ajax=1&full=1&amp;token=' . $this->getTokenForCron() . '&amp;redirect=1' . (Shop::getContext() == Shop::CONTEXT_SHOP ? '&id_shop=' . (int) Context::getContext()->shop->id : '') . '" class="btn-link">
> ```

Also here
___
*[controllers/admin/AdminSearchController.php, line 492 at r3](https://reviewable.io/reviews/prestashop/prestashop/10131#-LLdlvCh2oai7jN_GSsJ:-LVDrAmK2UuZ_KRITBhF:b-fcuexd) ([raw file](https://github.com/prestashop/prestashop/blob/a7824d1fe134196ca728ad38c5519d7e24487a69/controllers/admin/AdminSearchController.php#L492)):*
<details><summary><i>Previously, tomlev (Thomas LEVIANDIER) wrote…</i></summary><blockquote>

why 2 hours ? would you please provide some reason in comments to help understanding this arbitrary duration ?
</blockquote></details>

In any case, please set this number as a constant, then document the constant.
___
*[src/PrestaShopBundle/Controller/Admin/CommonController.php, line 68 at r4](https://reviewable.io/reviews/prestashop/prestashop/10131#-LKgLbidp0fJF9AOABCb-r4-68:-LVDrfMw0dso_eyS5uc_:b-c6ah) ([raw file](https://github.com/prestashop/prestashop/blob/bd4a26adadb6bd1b035073fe3b3631bf73548ead/src/PrestaShopBundle/Controller/Admin/CommonController.php#L68)):*
<details><summary><i>Previously, matks (Mathieu Ferment) wrote…</i></summary><blockquote>

Why us 😛 everybody can write an adapter
</blockquote></details>

Use a Query (CQRS)
___
*[src/PrestaShopBundle/Controller/Admin/CommonController.php, line 57 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDrYxk5WcDWJ0ginCG:-LVDrYxk5WcDWJ0ginCH:b-c6ah) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/src/PrestaShopBundle/Controller/Admin/CommonController.php#L57)):*
> ```PHP
>     {
>         // Adapter needed here?
>         return new JsonResponse((new \Notification())->getLastElements());
> ```

Use a Query (CQRS)
___
*[src/PrestaShopBundle/Form/Admin/Product/ProductInformation.php, line 146 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDsEwL1lLexDcZczw8:-LVDsEwL1lLexDcZczw9:bbintu8) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/src/PrestaShopBundle/Form/Admin/Product/ProductInformation.php#L146)):*
> ```PHP
>         ])
>             ->add('inputPackItems', TypeaheadProductPackCollectionType::class, [
>                 'remote_url' => $this->context->getAdminLink('AdminProducts') . '&ajax=1&action=productsList&forceJson=1&excludeVirtuals=1&limit=20&q=%QUERY',
> ```

Remember to use the 4th argument of getAdminLink
___
*[src/PrestaShopBundle/Form/Admin/Product/ProductInformation.php, line 288 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDsH4c-B7Rrm6jmS6Z:-LVDsH4c-B7Rrm6jmS6_:buo9nmt) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/src/PrestaShopBundle/Form/Admin/Product/ProductInformation.php#L288)):*
> ```PHP
>             ])
>             ->add('related_products', TypeaheadProductCollectionType::class, [
>                 'remote_url' => $this->context->getAdminLink('AdminProducts') . '&ajax=1&action=productsList&forceJson=1&disableCombination=1&exclude_packs=0&excludeVirtuals=0&limit=20&q=%QUERY',
> ```

Same here
___
*[src/PrestaShopBundle/Form/Admin/Product/ProductSeo.php, line 68 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDsJ6Q1tDlpHYkB8MO:-LVDsJ6Q1tDlpHYkB8MP:bbintu8) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/src/PrestaShopBundle/Form/Admin/Product/ProductSeo.php#L68)):*
> ```PHP
>     {
>         $remoteUrls = [
>             ProductInterface::REDIRECT_TYPE_PRODUCT_MOVED_PERMANENTLY => $this->context->getAdminLink('AdminProducts') . '&ajax=1&action=productsList&forceJson=1&disableCombination=1&exclude_packs=0&excludeVirtuals=0&limit=20&q=%QUERY',
> ```

Remember to use the 4th argument of getAdminLink
___
*[src/PrestaShopBundle/Form/Admin/Product/ProductSeo.php, line 69 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDsKEg12jUw85aYryc:-LVDsKEg12jUw85aYryd:buo9nmt) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/src/PrestaShopBundle/Form/Admin/Product/ProductSeo.php#L69)):*
> ```PHP
>         $remoteUrls = [
>             ProductInterface::REDIRECT_TYPE_PRODUCT_MOVED_PERMANENTLY => $this->context->getAdminLink('AdminProducts') . '&ajax=1&action=productsList&forceJson=1&disableCombination=1&exclude_packs=0&excludeVirtuals=0&limit=20&q=%QUERY',
>             ProductInterface::REDIRECT_TYPE_PRODUCT_FOUND => $this->context->getAdminLink('AdminProducts') . '&ajax=1&action=productsList&forceJson=1&disableCombination=1&exclude_packs=0&excludeVirtuals=0&limit=20&q=%QUERY',
> ```

Same here
___
*[src/PrestaShopBundle/Form/Admin/Product/ProductSeo.php, line 172 at r5](https://reviewable.io/reviews/prestashop/prestashop/10131#-LVDsKx837he7cIimbm6:-LVDsKx837he7cIimbm7:bkkjs1q) ([raw file](https://github.com/prestashop/prestashop/blob/cd33c012c894c9a0ce8d08b5ea758f734999ea70/src/PrestaShopBundle/Form/Admin/Product/ProductSeo.php#L172)):*
> ```PHP
>                 TypeaheadProductCollectionType::class,
>                 [
>                     'remote_url' => $this->context->getAdminLink('AdminProducts') . '&ajax=1&action=productsList&forceJson=1&disableCombination=1&exclude_packs=0&excludeVirtuals=0&limit=20&q=%QUERY',
> ```

And here

**Comment by Quetzacoalt91** [comment]
Ouch, regarding the latest comments, the PR seems to be now obsolete compared to the new features brought to the core.

This PR will be redone.

**Comment by tomlev** [comment]
hi @Quetzacoalt91 
you can use this commit to test ajax.php
https://github.com/tomlev/PrestaShop/commit/efa62aa8dbe472ed627feb4c46ccdc2a30bba88a

xxx

**Pull request closed without merging**